spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,298 @@
+    return StreamingOffset.fromJson(json);
+  }
+
+  @Override
+  public void commit(Offset end) {
+    // Since all the data and metadata of Iceberg is as it is, nothing needs to commit when
+    // offset is processed, so no need to implement this method.
+  }
+
+  @Override
+  public void stop() {}
+
+  @Override
+  @SuppressWarnings("unchecked")
+  protected List<CombinedScanTask> tasks() {
+    LOG.info("Processing data from {} exclusive to {} inclusive", startOffset, endOffset);
+
+    if (startOffset.equals(StreamingOffset.START_OFFSET) || endOffset.equals(StreamingOffset.START_OFFSET)) {
+      return Collections.emptyList();
+    }
+    if (startOffset.equals(endOffset)) {
+      return Collections.emptyList();
+    }
+
+    CloseableIterable<IndexedTask> tasks =
+        getChanges(startOffset.snapshotId(), startOffset.index(), startOffset.isStartingSnapshotId());
+    List<IndexedTask> pendingTasks = Lists.newArrayList();
+    for (IndexedTask t : tasks) {
+      if (t.snapshotId() == endOffset.snapshotId()) {
+        pendingTasks.add(t);
+        break;
+      } else {
+        pendingTasks.add(t);
+      }
+    }
+
+    CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(
+        CloseableIterable.combine(pendingTasks.stream().map(IndexedTask::task).collect(Collectors.toList()), tasks),
+        splitSize);
+    return Lists.newArrayList(
+        TableScanUtil.planTasks(splitTasks, splitSize, lookback, openFileCost));
+  }
+
+  private StreamingOffset calculateStartingOffset() {
+    StreamingOffset startingOffset;
+    if (startSnapshotId != null) {
+      startingOffset = new StreamingOffset(startSnapshotId, -1, true);
+    } else {
+      List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+      if (snapshotIds.isEmpty()) {
+        // there's no snapshot currently.
+        startingOffset = StreamingOffset.START_OFFSET;
+      } else {
+        startingOffset = new StreamingOffset(snapshotIds.get(snapshotIds.size() - 1), -1, true);
+      }
+    }
+
+    return startingOffset;
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    CloseableIterable<IndexedTask> pendingTasks = getChanges(start.snapshotId(), start.index(),
+        start.isStartingSnapshotId());
+    CloseableIterable<IndexedTask> rateLimitedTasks = rateLimit(pendingTasks, maxFilesPerTrigger);
+    IndexedTask last = Iterables.size(rateLimitedTasks) == 0 ? null : Iterables.getLast(rateLimitedTasks);
+
+    if (last == null) {
+      return start;
+    } else {
+      boolean isStarting = last.snapshotId() == start.snapshotId() && start.isStartingSnapshotId();
+      return new StreamingOffset(last.snapshotId(), last.index(), isStarting);
+    }
+  }
+
+  @VisibleForTesting
+  CloseableIterable<IndexedTask> getChanges(long snapshotId, int index, boolean isStarting) {
```

Comment:
`getChanges` is an expensive method call because it may result in planning a full scan of the input table. And even if it isn't a starting snapshot, we expect this to scan manifest files for append commits.

Because this is an expensive method, I think this streaming reader should avoid calling it multiple times. I think this should get the result of `getChanges` and store it in a field so that the same iterable can be used across batches.

This should also consume changes just once. Instead of calling this to find where the end offset should be and then calling this to get tasks between the start and end offsets, I think this reader should accumulate a list of tasks in `calculateEndOffset`, and store those tasks as an instance field. When `tasks` is called to fetch the tasks, it should validate that the end offset hasn't changed and return that list of tasks.

Reply:
Yeah, I was thinking about changing to this way, I will take a shot on this.

Reply:
Changed.